sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/ByteKeyRangeTracker.java

```diff
@@ -55,13 +60,23 @@ public synchronized ByteKeyRange getRange() {
 
   @Override
   public synchronized boolean tryReturnRecordAt(boolean isAtSplitPoint, ByteKey recordStart) {
+    if (position == null && !isAtSplitPoint) {
+      throw new IllegalStateException(
+          String.format("The first record [starting at %d] must be at a split point", recordStart));
+    }
     if (isAtSplitPoint && !range.containsKey(recordStart)) {
+      if (recordStart.compareTo(range.getEndKey()) >= 0) {
+        done = true;
+      }
```

Comment:
I think the check at L68-70 does not belong. Should always become `done` when we return `false`